### ISSUE TITLE
hvd.shutdown() hangs in MPI_Finalize() for just one rank

### ISSUE THREAD

**Issue by maxhgerlach** [horovod/horovod#1102]
**Environment:**
1. Framework: TensorFlow
2. Framework version: 1.13.1, custom built without XLA
3. Horovod version: 0.16.2
4. MPI version: Open MPI 4.0.1
5. CUDA version: 10.0
6. NCCL version: 2.3.7-1+cuda10.0
7. Python version: 2.7
8. OS and version: Ubuntu 16.04.5 LTS

Mellanox OFED 4.4-2.0.7.0 is installed.

**Bug report:**
I am facing an issue in our custom Horovod-based training code. When I call `hvd.shutdown()` on all ranks, one of the ranks hangs and does not return. If I do not call `hvd.shutdown()` explicitly, I observe the same problem when it is called by the atexit handler.

To pinpoint the problem I added extra tracing log messages in BackgroundThreadLoop (https://github.com/maxhgerlach/horovod/commits/maxg-shutdown-tracing): It hangs inside `MPI_Finalize()` for that one rank. In this [example log](https://gist.github.com/maxhgerlach/2c7030198b86da5093c56272ede114b6) rank 0 never moves past `MPI_Finalize()`, but in other test runs some other rank hangs.

This is with Open MPI 4.0.1; previously I used version 3.1.2. With the older version, the problematic rank would not hang in `MPI_Finalize`, but abort the program with an exit code of 244. With either version 3.1.2 or version 4.0.1 setting a different error handler via `MPI_Comm_set_errhandler(MPI_COMM_WORLD, MPI_ERRORS_RETURN);` did not change the observed behaviour. (Supposedly this would cause any MPI routine to actually return a nonzero code on error, rather than directly terminate the whole program.)

Inserting an mpi4py `MPI.COMM_WORLD.barrier()` right before `hvd.shutdown()` makes no difference either.

I am at a loss how to debug this.

**Comment by alsrgv** [comment]
@maxhgerlach, thanks for reporting this issue.  What the minimum number of processes does it take to reproduce it?  Would it fail with two processes on a single node, or do they have to be on different nodes?

**Comment by maxhgerlach** [comment]
@alsrgv, thanks for looking into this.

The log above was for 8 processes on the same host.

I was able to reproduce the hanging shutdown issue already with 3 processes on the same host. With only 2 processes there was no issue.

Interestingly, with 3 processes there are some extra error messages near initialization:
```
--------------------------------------------------------------------------
It looks like orte_init failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during orte_init; some of which are due to configuration or
environment problems.  This failure appears to be an internal failure;
here's some additional information (which may only be relevant to an
Open MPI developer):

  getting local rank failed
  --> Returned value No permission (-17) instead of ORTE_SUCCESS
--------------------------------------------------------------------------
--------------------------------------------------------------------------
It looks like orte_init failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during orte_init; some of which are due to configuration or
environment problems.  This failure appears to be an internal failure;
here's some additional information (which may only be relevant to an
Open MPI developer):

  orte_ess_init failed
  --> Returned value No permission (-17) instead of ORTE_SUCCESS
--------------------------------------------------------------------------
--------------------------------------------------------------------------
It looks like MPI_INIT failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during MPI_INIT; some of which are due to configuration or environment
problems.  This failure appears to be an internal failure; here's some
additional information (which may only be relevant to an Open MPI
developer):

  ompi_mpi_init: ompi_rte_init failed
  --> Returned "No permission" (-17) instead of "Success" (0)
--------------------------------------------------------------------------
```
But training proceeds until shutdown, where it hangs. Could this be related?

This is how I launch (I haven't looked into horovodrun yet):
`mpirun --prefix /opt/openmpi -np 3 -H localhost:3 -bind-to none -map-by slot -x NCCL_DEBUG=INFO -x NCCL_DEBUG_FILE=/dev/stderr -x LD_LIBRARY_PATH -x PATH -mca pml ob1 -mca btl ^openib --tag-output --timestamp-output python [our script and arguments]`

**Comment by maxhgerlach** [comment]
We use multiprocessing inside our training script to fork out some IO work. Could this be a problem? So the process tree for the job with one rank hanging in MPI_finalize looks like this:
```
$ ps fux
   PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
[...]   
231625  0.0  0.0 377152 17256 ?        Sl   15:58   0:00                  \_ mpirun --prefix /opt/openmpi -np 4 -H localhost:4 -bind-to none -map-by slot -x NCCL_DEBUG=INFO -x NCCL_DEBUG_F
231633 11.0  2.0 41385640 3956868 ?    Sl   15:58   5:07                      \_ python [training script]
232676  5.0  1.4 40055256 2848648 ?    Sl   15:59   2:17                          \_ python [training script]
[...]
```
I have attached gdb to both processes and collected backtraces for all threads: 
1. [backtraces for 231633, 75 threads](https://gist.github.com/maxhgerlach/3e7ce94d3d4f6667d876561aacdb96e0)
2. [backtraces for 232676, 2 threads](https://gist.github.com/maxhgerlach/627fe7383dbcb70299ab5f3d08915d7d)

**Comment by alsrgv** [comment]
@maxhgerlach, forking could be an issue. Does your code have a mode where it could run w/o forked process and IO work, just to see if this resolves the issue?  You could potentially use multithreading to avoid forking.

**Comment by maxhgerlach** [comment]
Hmm, I changed our code such that the forked-out multiprocessing process is definitely terminated before hvd.shutdown() is called [it was started after hvd.init() was called in the master process]. Now the issue is much harder to reproduce, but I still saw it once in 16 trials. 

Unfortunately it would be quite tricky to get completely rid of the separate process. Threading slowed down training too much because of the GIL.

**Comment by maxhgerlach** [comment]
@alsrgv: Could there still be issues from forking, if I fork _before_ calling `hvd.init()` and never call any MPI routine in the forked-out process?

**Comment by alsrgv** [comment]
@maxhgerlach, less likely, but forking can make MPI get quite crazy due to all the malloc hooks and madvising (for RDMA) that it does.

If you could modify the program to run w/o forking just to test if it hangs (perhaps replace real data fetch with synthetic data?), we could confirm whether forking is causing the issue in the first place and go from there.

**Comment by maxhgerlach** [comment]
@alsrgv, last week I have modified our program such that the fork happens earlier than hvd.init(). Since then I could not observe the issue anymore. I am running some tests periodically that triggered the problem before. If it resurfaces, I will be happy to devise a test case with a synthetic or synchronous data pipeline.

Do you believe it would be a good idea to mention in the Horovod docs that Python-style multiprocessing could cause some grief?

**Comment by alsrgv** [comment]
@maxhgerlach, thanks for the report!  I believe it'll be good to add to the docs.
